dockershim: use the new Metadata types and add more unit tests

yujuhong opened the pull request:
Part of #31459

<!-- Reviewable:start -->

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31573)

<!-- Reviewable:end -->

yujuhong committed:
dockershim: bug fixes and more unit tests

Fixing the name triming and other small bugs. Added sandbox listing unit tests.

yujuhong committed:
dockershim: move naming helpers to a separate file

yujuhong committed:
dockershim: utilize the Metadata in container names

This commit changes how the shim constructs and parses docker container names
by using the new "Metadata" types.

feiskyer commented on pkg/kubelet/dockershim/convert.go:
```diff
@@ -49,20 +49,20 @@ func toRuntimeAPIImage(image *dockertypes.Image) (*runtimeApi.Image, error) {
 		Image:    &runtimeApi.ImageSpec{Image: &c.Image},
 		ImageRef: &c.ImageID,
 		State:    &state,
 		Labels:   c.Labels,
```
nit: add a todo for extracting annotations from labels.

feiskyer commented on pkg/kubelet/dockershim/docker_sandbox.go:
```diff
@@ -182,19 +178,24 @@ func (ds *dockerService) ListPodSandbox(filter *runtimeApi.PodSandboxFilter) ([]
-		if filterOutReadySandboxes && s.GetState() == runtimeApi.PodSandBoxState_READY {
+		converted, err := toRuntimeAPISandbox(&c)
+		if err != nil {
+			glog.V(5).Infof("Unable to convert docker to runtime API container: %v", err)
```
nit: replace `container` with `sandbox`.

feiskyer commented on pkg/kubelet/dockershim/naming.go:
```diff
@@ -0,0 +1,125 @@
+	name = strings.TrimPrefix(name, "/")
+
+	parts := strings.Split(name, nameDelimiter)
+	if len(parts) != 6 {
```
nit: also checks whether the first part is equal to kubePrefix here.

feiskyer commented on pkg/kubelet/dockershim/naming.go:
```diff
@@ -0,0 +1,125 @@
+	name = strings.TrimPrefix(name, "/")
+
+	parts := strings.Split(name, nameDelimiter)
+	if len(parts) != 6 {
```
nit: also checks whether the first part is equal to kubePrefix here.

yujuhong commented:
Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/kubelet/dockershim/convert.go, line 64 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31573#-KQ8klRHHEEL1fOFjSa1-r3-64:-KQMC6Bz8FIYSgCqQZqW:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9465044db04b52d9bbdaf7c5f16da7190a69420a/pkg/kubelet/dockershim/convert.go#L64)):*
<details><summary><i>Previously, feiskyer (Pengfei Ni) wrote…</i></summary>
> nit: add a todo for extracting annotations from labels.
</details>
Done.

---

*[pkg/kubelet/dockershim/docker_sandbox.go, line 186 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31573#-KQ8klRFyj9anfiLuoAm-r3-186:-KQMCHPX-L6zsrPHgy8o:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9465044db04b52d9bbdaf7c5f16da7190a69420a/pkg/kubelet/dockershim/docker_sandbox.go#L186)):*
<details><summary><i>Previously, feiskyer (Pengfei Ni) wrote…</i></summary>
> nit: replace `container` with `sandbox`.
</details>
Done.

---

*[pkg/kubelet/dockershim/naming.go, line 89 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31573#-KQ8klRGzXA_jJc4JNH--r3-89:-KQMG8wXx6MEYy8DiE5b:bn21kf1) ([raw file](https://github.com/kubernetes/kubernetes/blob/9465044db04b52d9bbdaf7c5f16da7190a69420a/pkg/kubelet/dockershim/naming.go#L89)):*
<details><summary><i>Previously, feiskyer (Pengfei Ni) wrote…</i></summary>
> nit: also checks whether the first part is equal to kubePrefix here. 
</details>
Done. Also added a test.

---

*[pkg/kubelet/dockershim/naming.go, line 112 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31573#-KQ8klRGzXA_jJc4JNH--r3-112:-KQMGBt_9XWYfAHR4tOn:bn21kf1) ([raw file](https://github.com/kubernetes/kubernetes/blob/9465044db04b52d9bbdaf7c5f16da7190a69420a/pkg/kubelet/dockershim/naming.go#L112)):*
<details><summary><i>Previously, feiskyer (Pengfei Ni) wrote…</i></summary>
> nit: also checks whether the first part is equal to kubePrefix here. 
</details>
Done. Also added a test.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31573)*
<!-- Sent from Reviewable.io -->

yujuhong commented:
Duh...I forgot to include a unit test file (`naming_test.go`). PTAL again. Thanks!

feiskyer commented:
LGTM.

euank commented:
LGTM as well

yujuhong commented:
Rebased and re-applied lgtm
